services/wallet/history/service.go

```diff
@@ -0,0 +1,339 @@
+		Type: eventType,
+		Accounts: []common.Address{
+			common.Address(account),
+		},
+		Message: message,
+	})
+}
+
+func (s *Service) StartBalanceHistory() {
+	go func() {
+		s.serviceContext, s.cancelFn = context.WithCancel(context.Background())
+		s.timer = time.NewTimer(balanceHistoryUpdateInterval)
+
+		update := func() (exit bool) {
+			err := s.updateBalanceHistoryForAllEnabledNetworks(s.serviceContext)
+			if s.serviceContext.Err() != nil {
+				s.triggerEvent(EventBalanceHistoryUpdateFinished, statustypes.Address{}, "Service canceled")
+				s.timer.Stop()
+				return true
+			}
+			if err != nil {
+				s.triggerEvent(EventBalanceHistoryUpdateFinishedWithError, statustypes.Address{}, err.Error())
+			}
+			return false
+		}
+
+		if update() {
+			return
+		}
+
+		for range s.timer.C {
+			resetTimer(s.timer, balanceHistoryUpdateInterval)
+
+			if update() {
+				return
+			}
+		}
+	}()
+}
+
+func resetTimer(timer *time.Timer, interval time.Duration) {
+	if timer != nil {
+		timer.Stop()
+		timer.Reset(interval)
+	}
+}
+
+func (s *Service) UpdateVisibleTokens(symbols []string) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	startUpdate := len(s.visibleSymbols) == 0 && len(symbols) > 0
+	s.visibleSymbols = symbols
+	if startUpdate {
+		resetTimer(s.timer, 0)
+	}
+}
+
+func (s *Service) isTokenVisible(symbol string) bool {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	for _, visibleSymbol := range s.visibleSymbols {
+		if visibleSymbol == symbol {
+			return true
+		}
+	}
+	return false
+}
+
+// Native token implementation of DataSource interface
+type chainClientSource struct {
+	chainClient *chain.Client
+	currency    string
+}
+
+func (src *chainClientSource) HeaderByNumber(ctx context.Context, blockNo *big.Int) (*types.Header, error) {
+	return src.chainClient.HeaderByNumber(ctx, blockNo)
+}
+
+func (src *chainClientSource) BalanceAt(ctx context.Context, account common.Address, blockNo *big.Int) (*big.Int, error) {
+	return src.chainClient.BalanceAt(ctx, account, blockNo)
+}
+
+func (src *chainClientSource) ChainID() uint64 {
+	return src.chainClient.ChainID
+}
+
+func (src *chainClientSource) Currency() string {
+	return src.currency
+}
+
+func (src *chainClientSource) TimeNow() int64 {
+	return time.Now().UTC().Unix()
+}
+
+type tokenChainClientSource struct {
+	chainClientSource
+	TokenManager   *token.Manager
+	NetworkManager *network.Manager
+}
+
+func (src *tokenChainClientSource) BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error) {
+	network := src.NetworkManager.Find(src.chainClient.ChainID)
+	if network == nil {
+		return nil, errors.New("network not found")
+	}
+	token := src.TokenManager.FindToken(network, src.currency)
+	if token == nil {
+		return nil, errors.New("token not found")
+	}
+	balance, err := src.TokenManager.GetTokenBalanceAt(ctx, src.chainClient, account, token.Address, blockNumber)
+	if err != nil {
+		if err.Error() == "no contract code at given address" {
```

Comment:
Could we not reuse that to indicate that we should not continue to go back in time?

Reply:
I added `firstUnavailableBlockNo` to mark the oldest known `blockNumber` without contract and avoid fetching this later. Otherwise, it is needed to mark history cached. No need for a specific state, a balance of `0` works best.